File: ts-sdk/src/routines.ts
/**
 * Routines API - scheduled prompts for agents.
 */
import { requestWithRetry, responseAPIError } from './http.js';
import { getAgentsApiBaseUrl } from './config.js';

function envValue(key: string): string | undefined {
  const maybeProcess = globalThis as unknown as { process?: { env?: Record<string, string | undefined> } };
  return maybeProcess.process?.env?.[key];
}

export function deriveRoutinesApiBase(agentsApiBase?: string): string {
  const configured = envValue('HYPER_ROUTINES_API_BASE');
  const raw = (configured || agentsApiBase || getAgentsApiBaseUrl()).replace(/\/$/, '');
  const url = new URL(raw.includes('://') ? raw : `https://${raw}`);
  let path = url.pathname.replace(/\/$/, '');
  if (path.endsWith('/routines')) {
    return `${url.protocol}//${url.host}${path}`;
  }
  if (path.endsWith('/agents')) {
    path = path.slice(0, -'/agents'.length);
  }
  return `${url.protocol}//${url.host}${path}/routines`;
}

export interface Routine {
  id: string;
  userId: string;
  agentId: string;
  cron: string | null;
  prompt: string;
  enabled: boolean;
  name: string | null;
  runAt: string | null;
  nextRunAt: string | null;
  createdAt: string | null;
  updatedAt: string | null;
}

export interface RoutineCreateOptions {
  agentId: string;
  prompt: string;
  cron?: string;
  runAt?: string;
  name?: string;
  enabled?: boolean;
}

export interface RoutineUpdateOptions {
  agentId?: string;
  cron?: string;
  prompt?: string;
  enabled?: boolean;
  name?: string;
}

function routineFromDict(data: any): Routine {
  return {
    id: String(data?.id || ''),
    userId: String(data?.user_id || data?.userId || ''),
    agentId: String(data?.agent_id || data?.agentId || ''),
    cron: data?.cron ?? null,
    prompt: data?.prompt || '',
    enabled: Boolean(data?.enabled ?? false),
    name: data?.name ?? null,
    runAt: data?.run_at ?? data?.runAt ?? null,
    nextRunAt: data?.next_run_at ?? data?.nextRunAt ?? null,
    createdAt: data?.created_at ?? data?.createdAt ?? null,
    updatedAt: data?.updated_at ?? data?.updatedAt ?? null,
  };
}

async function handleResponse<T = any>(response: Response, method?: string): Promise<T> {
  if (response.status >= 400) {
    throw await responseAPIError(response, method);
  }
  if (response.status === 204 || response.status === 205) return undefined as T;
  const text = await response.text();
  if (!text) return undefined as T;
  return JSON.parse(text) as T;
}

function encodeRef(value: string): string {
  return encodeURIComponent(value);
}

export class RoutinesAPI {
  private apiBase: string;
  private apiKey: string;
  private timeout: number;

  constructor(apiKey: string, options: { apiBase?: string; agentsApiBase?: string; timeout?: number } = {}) {
    if (!apiKey) {
      throw new Error('API key required for routines');
    }
    this.apiKey = apiKey;
    this.apiBase = (options.apiBase || deriveRoutinesApiBase(options.agentsApiBase)).replace(/\/$/, '');
    this.timeout = options.timeout ?? 30000;
  }

  private headers(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json',
    };
  }

  private async request<T = any>(
    method: string,
    path: string,
    body?: any,
    params?: Record<string, string>,
  ): Promise<T> {
    let url = `${this.apiBase}${path}`;
    if (params) {
      const query = new URLSearchParams(params).toString();
      if (query) url = `${url}?${query}`;
    }
    const response = await requestWithRetry({
      method,
      url,
      headers: this.headers(),
      body,
      retries: method === 'GET' ? 3 : 1,
      timeout: this.timeout,
    });
    return handleResponse<T>(response, method);
  }

  async list(options: { agentId?: string } = {}): Promise<Routine[]> {
    const params = options.agentId !== undefined ? { agent_id: options.agentId } : undefined;
    const data = await this.request<any[]>('GET', '', undefined, params);
    if (!Array.isArray(data)) {
      throw new Error('Routines response must be an array.');
    }
    return data.map(routineFromDict);
  }

  async get(routineId: string): Promise<Routine> {
    const data = await this.request('GET', `/${encodeRef(routineId)}`);
    return routineFromDict(data);
  }

  async create(body: RoutineCreateOptions): Promise<Routine> {
    const payload: Record<string, unknown> = { agent_id: body.agentId };
    if (body.cron !== undefined) payload.cron = body.cron;
    payload.prompt = body.prompt;
    payload.enabled = body.enabled ?? true;
    if (body.runAt !== undefined) payload.run_at = body.runAt;
    if (body.name !== undefined) payload.name = body.name;
    const data = await this.request('POST', '', payload);
    return routineFromDict(data);
  }

  async update(routineId: string, body: RoutineUpdateOptions): Promise<Routine> {
    const payload: Record<string, unknown> = {};
    if (body.agentId !== undefined) payload.agent_id = body.agentId;
    if (body.cron !== undefined) payload.cron = body.cron;
    if (body.prompt !== undefined) payload.prompt = body.prompt;
    if (body.enabled !== undefined) payload.enabled = body.enabled;
    if (body.name !== undefined) payload.name = body.name;
    const data = await this.request('PATCH', `/${encodeRef(routineId)}`, payload);
    return routineFromDict(data);
  }

  async delete(routineId: string): Promise<void> {
    await this.request('DELETE', `/${encodeRef(routineId)}`);
  }
}

File: sdk/hypercli/routines.py
"""Routines API client."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from urllib.parse import quote, urlsplit

from .config import get_agents_api_base_url, get_config_value
from .workspaces import _request


def _derive_routines_base(agents_api_base: str | None = None) -> str:
    configured = get_config_value("HYPER_ROUTINES_API_BASE")
    if configured:
        raw = configured.strip().rstrip("/")
    else:
        raw = (agents_api_base or get_agents_api_base_url()).strip().rstrip("/")
    parsed = urlsplit(raw if "://" in raw else f"https://{raw}")
    path = parsed.path.rstrip("/")
    if path.endswith("/routines"):
        return f"{parsed.scheme}://{parsed.netloc}{path}"
    if path.endswith("/agents"):
        path = path[: -len("/agents")]
    return f"{parsed.scheme}://{parsed.netloc}{path}/routines"


def _encode_ref(value: str) -> str:
    return quote(value, safe="")


def _parse_datetime(value) -> datetime | None:
    if value is None or isinstance(value, datetime):
        return value
    if not isinstance(value, str):
        raise ValueError(f"Invalid routine timestamp: {value}")
    timestamp = datetime.fromisoformat(value.replace("Z", "+00:00"))
    if timestamp.tzinfo is None:
        timestamp = timestamp.replace(tzinfo=timezone.utc)
    return timestamp


@dataclass
class Routine:
    id: str
    user_id: str
    agent_id: str
    cron: str | None
    prompt: str
    enabled: bool
    name: str | None = None
    run_at: str | None = None
    next_run_at: datetime | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None

    @classmethod
    def from_dict(cls, data: dict) -> "Routine":
        return cls(
            id=str(data.get("id", "")),
            user_id=str(data.get("user_id", "")),
            agent_id=str(data.get("agent_id", "")),
            cron=data.get("cron"),
            prompt=data.get("prompt", ""),
            enabled=bool(data.get("enabled", True)),
            name=data.get("name"),
            run_at=data.get("run_at"),
            next_run_at=_parse_datetime(data.get("next_run_at")),
            created_at=_parse_datetime(data.get("created_at")),
            updated_at=_parse_datetime(data.get("updated_at")),
        )


class RoutinesAPI:
    """Client for the routines service mounted at /routines."""

    def __init__(self, api_key: str, api_base: str | None = None, agents_api_base: str | None = None):
        if not api_key:
            raise ValueError("API key required for routines")
        self.api_key = api_key
        self.api_base = (api_base or _derive_routines_base(agents_api_base)).rstrip("/")

    def list(self, *, agent_id: str | None = None) -> list[Routine]:
        params = {"agent_id": agent_id} if agent_id else None
        data = _request("GET", self.api_base, api_key=self.api_key, params=params)
        return [Routine.from_dict(item) for item in data]

    def get(self, routine_id: str) -> Routine:
        data = _request(
            "GET",
            f"{self.api_base}/{_encode_ref(routine_id)}",
            api_key=self.api_key,
        )
        return Routine.from_dict(data)

    def create(
        self,
        *,
        agent_id: str,
        prompt: str,
        cron: str | None = None,
        run_at: str | None = None,
        name: str | None = None,
        enabled: bool = True,
    ) -> Routine:
        payload = {
            "agent_id": agent_id,
            "prompt": prompt,
            "enabled": enabled,
        }
        if cron is not None:
            payload["cron"] = cron
        if run_at is not None:
            payload["run_at"] = run_at
        if name is not None:
            payload["name"] = name
        data = _request("POST", self.api_base, api_key=self.api_key, json=payload)
        return Routine.from_dict(data)

    def update(
        self,
        routine_id: str,
        *,
        agent_id: str | None = None,
        cron: str | None = None,
        prompt: str | None = None,
        enabled: bool | None = None,
        name: str | None = None,
    ) -> Routine:
        payload = {}
        if agent_id is not None:
            payload["agent_id"] = agent_id
        if cron is not None:
            payload["cron"] = cron
        if name is not None:
            payload["name"] = name
        if prompt is not None:
            payload["prompt"] = prompt
        if enabled is not None:
            payload["enabled"] = enabled
        data = _request(
            "PATCH",
            f"{self.api_base}/{_encode_ref(routine_id)}",
            api_key=self.api_key,
            json=payload,
        )
        return Routine.from_dict(data)

    def delete(self, routine_id: str) -> dict:
        return _request(
            "DELETE",
            f"{self.api_base}/{_encode_ref(routine_id)}",
            api_key=self.api_key,
        )
